packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -386,6 +403,47 @@ export class Pipeline extends Resource implements IPipeline {
     };
   }
 
+  private attachActionForCrossAccount(action: Action): iam.IRole | undefined {
+    if (!action.resource) {
+      return action.role;
+    }
+
+    const pipelineStack = this.node.stack;
+    const resourceStack = action.resource.node.stack;
+    if (pipelineStack.env.account !== resourceStack.env.account) {
+      if (!this.artifactBucket.encryptionKey) {
+        throw new Error('The Pipeline is being used in a cross-account manner, ' +
+            'but its artifact Bucket does not have a KMS Key defined. ' +
+            'A KMS key is required for a cross-account Pipeline. ' +
+            'Make sure to pass a Bucket with a Key when creating the Pipeline');
+      }
+
+      if (action.role) {
+        return action.role;
+      }
+
+      let actionRole = this.crossAccountActionRoles.get(resourceStack);
+      if (!actionRole) {
+        actionRole = new iam.Role(resourceStack, this.node.id + 'ActionRole', {
```

Comment:
Should this also be `this.node.uniqueId` instead of just `this.node.id`?

Reply:
Hmmm... on second thought, these are per-resource Stack unique, and the Pipeline ID makes it per-Pipeline unique... maybe leave it as `this.node.id`?